internal/iterator/fan_in.go

```diff
@@ -10,46 +10,63 @@ import (
 )
 
 type FanIn struct {
-	ctx       context.Context
-	cancel    context.CancelFunc
-	count     int
-	out       chan *Msg
-	addCh     chan (chan *Msg)
-	mu        sync.Mutex
-	accepting bool
-	pool      *pool.ContextPool
-	drained   chan bool
+	ctx        context.Context
+	cancel     context.CancelFunc
+	out        chan *Msg
+	addCh      chan (chan *Msg)
+	drainQueue []chan *Msg
+	accepting  bool
+	mu         sync.Mutex
+	wg         sync.WaitGroup
+	pool       *pool.ContextPool
 }
 
 func NewFanIn(ctx context.Context, limit int) *FanIn {
 	ctx, cancel := context.WithCancel(ctx)
 
 	f := &FanIn{
-		ctx:       ctx,
-		cancel:    cancel,
-		count:     0,
-		out:       make(chan *Msg, limit),
-		addCh:     make(chan (chan *Msg), limit),
-		accepting: true,
-		mu:        sync.Mutex{},
-		pool:      concurrency.NewPool(ctx, limit),
-		drained:   make(chan bool, 1),
+		ctx:        ctx,
+		cancel:     cancel,
+		out:        make(chan *Msg, limit),
+		addCh:      make(chan (chan *Msg), limit),
+		drainQueue: make([]chan *Msg, 0),
+		accepting:  true,
+		mu:         sync.Mutex{},
+		wg:         sync.WaitGroup{},
+		pool:       concurrency.NewPool(ctx, limit),
 	}
 
+	f.wg.Add(1)
 	go f.run()
+
 	return f
 }
 
+func (f *FanIn) cleaner() {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	for _, ch := range f.drainQueue {
+		Drain(ch).Wait()
+	}
+	f.wg.Done()
+}
+
 func (f *FanIn) run() {
 	defer func() {
 		f.Done()
 		_ = f.pool.Wait()
 		close(f.out)
 		for ch := range f.addCh {
-			drainOnExit(ch)
+			f.drainOnExit(ch)
 		}
-		f.drained <- true
-		close(f.drained)
+		f.mu.Lock()
+		queueSize := len(f.drainQueue)
+		f.mu.Unlock()
+		if queueSize > 0 {
+			f.wg.Add(1)
+			go f.cleaner()
+		}
+		f.wg.Done()
```

Comment:
Who is waiting for this wg?  Is this only for unit test purpose?  If so, can you add this as comment?

Reply:
Added comment.